### PR TITLE
Block Editor: Add controlled inner blocks focus styles.

### LINKED ISSUE #19886: Edit site: Template part focus styling
It should be obvious when a user starts editing a template part as these changes impact any page using the template part. I propose we add a focus state when editing a template part or its contents.

Note: My mockups are a frankenstein of low-fi mocks and some newer styles. They're not to be taken literally as much as convey an idea.

Here's a rough example to give you an idea of what I'm referring to:

image

Here's a [quick prototype of what I'm proposing](https://www.figma.com/proto/fud7MwfNd2rS0fjxMRdHwu/Full-Site-Editing-%E2%80%93-Templates?node-id=243%3A2360&viewport=-266%2C302%2C0.5&scaling=min-zoom).

And a big old gif of the prototype:

template-part-focus-i1

@epiqueras would you be interested in tackling this? I can help with styles.

### PR DESCRIPTION
**Pull request by epiqueras**
Closes #19886

## Screenshots

gif

## Types of Changes

*New Feature:* Controlled inner blocks like template parts and post content now have a distinct style when selected.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/block-editor/src/components/inner-blocks/index.js [hunk 1]
```diff
@@ -138,13 +138,17 @@ class InnerBlocks extends Component {
...
+			isSelected,
 			...props
 		} = this.props;
 		const { templateInProcess } = this.state;
 
 		const classes = classnames( 'block-editor-inner-blocks', {
 			'has-overlay': enableClickThrough && hasOverlay,
 			'is-capturing-toolbar': captureToolbars,
+			'is-controlled': onInput,
+			'is-selected': isSelected,
```
Comment (ellatrix) [inline packages/block-editor/src/components/inner-blocks/index.js]: Could we add these to `BlockList`? I'm trying to move the existing classes in #19849.

Comment (ellatrix) [reply packages/block-editor/src/components/inner-blocks/index.js]: That's because this particular wrapper will eventually be opt-out: #19910.

Comment (epiqueras) [reply packages/block-editor/src/components/inner-blocks/index.js]: Yeah, sure.

**Comment by shaunandrews** [comment]
This feels incredible similar to the current Spotlight mode, but with different styling. Maybe we should try fading all the other elements away, rather than introducing the dark overlay.

Fade Out

**Comment by epiqueras** [comment]
Feel free to tweak the shadow values.

**Comment by jasmussen** [comment]
You folks move so fast, it's wonderful to see!

This "focus mode"/isolation mode is in principle not a bad idea. Variations of it have been explored in a number of mockups, and the goal, nearly always is to:

- focus attention on where it should be
- in the case of template parts, make it clear changes are global

The 2nd one is probably the most important, and it affords the opportunity to add an "Are you sure" dialog when you exit editing the template part, and a spot to explain that — hey, this change affects every page! As such, the idea, the implementation, both feel like they are veering in the right direction. Kudos for the energy.

**I wonder though, whether this is the right time to implement it or not.** It’s not a bad idea, but it might be one that should be postponed until we know we need it. 

The G2 interface (#18667) is likely going to simplify a number of things, we might even consider retiring "spotlight mode" entirely, opening up alternate treatments to this, as Shaun alludes to.

More importantly, while we all THINK it's going to be important to make it ultra clear to users that they are editing a global part. But we don't actually **know for sure that this is a problem**. By distinguishing between a template part and any other block, we may in fact add confusion: **why is this block highlighted but not that block?** — and since we are already looking at batching multiple template part edits together in an on-publish prompt, we already have an opportunity to inform the users of what they are doing. 

To summarize: this may be the right idea and the right solution, but because we don't know that for sure yet, maybe we should wait!

**Comment by epiqueras** [comment]
>we don't actually know for sure yet that this is a problem.

Try editing a template part, it's very hard to tell if you are editing it or the blocks right after it without opening the block navigator.

>why is this block highlighted but not that block?

That can easily be explained somewhere like the breadcrumb or where the button to exit the code view appears.

**Comment by epiqueras** [comment]
Any updates on this from design?

**Comment by shaunandrews** [comment]
Here's a quick [mockup in Figma](https://www.figma.com/file/vngRayKCIrqcwXuogzNakn/Template-Partial-Selection?node-id=1%3A631):

Highlight Template Partial

**Comment by mapk** [comment]
That last one from @shaunandrews is spot on! I suggest that be the direction we go. All my efforts were leading in that direction as well. 

Let's just maintain the same amount of padding above the text within the selected template-part that initially shows in the unselected view. Once the template-part is focused, the padding above the text gets a little tight. Adding the margin for the block toolbar with the "pull away" effect is totally worth keeping. The breadcrumb indicator is unobtrusively informative as well. Really good work! 😍

**Comment by epiqueras** [comment]
gif

**Comment by epiqueras** [comment]
G2 makes it look slick @youknowriad :clap:

**Comment by noahtallen** [comment]
I was looking at picking this up again today. Should the "focus mode" styles be consistent with what we're doing here?

File: packages/block-editor/src/components/block-list/block.js [hunk 2]
```diff
@@ -121,6 +121,7 @@ function BlockListBlock( {
 			'is-focus-mode': isFocusMode,
 			'has-child-selected': isAncestorOfSelectedBlock,
 			'is-block-collapsed': isAligned,
+			'is-focus-container': name === 'core/template-part',
```
Comment (noahtallen) [inline packages/block-editor/src/components/block-list/block.js]: We need to create this class in a more extensible way. Before, with `InnerBlocks`, we added our custom styles if InnerBlocks had the `onInput` prop and if it or one of its children was selected.

**Comment by noahtallen** [comment]
This is what it looks like right now. It feels a little choppy because selecting blocks in the editor jerks everything around a bit (and because my gif software is low framerate!)

**Comment by jasmussen** [comment]
> I was looking at picking this up again today. Should the "focus mode" styles be consistent with what we're doing here?

Yes this is an excellent point, and I think we should be changing the style of the effect in this PR to match that of what focus mode does.

File: packages/block-editor/src/components/inner-blocks/index.js (hunk 1 above)
Comment (noahtallen) [reply packages/block-editor/src/components/inner-blocks/index.js]: I've moved these classes to block-list-block now. Does that change make sense??

**Comment by noahtallen** [comment]
Cool! That makes it easy to stay consistent if we did ever want to update focus mode to be similar to the mochup above.

**Comment by MichaelArestad** [comment]
I'm thinking we should probably shelve this for now in favor of just using the usual block focus styles.

**Comment by noahtallen** [comment]
Yeah, that sounds reasonable. I'll close this PR out since the styles basically just set a massive background shadow over the rest of the page. (And since it's quite stale at this point!) I investigated triggering the focus styles just for the template part block probably 4 weeks ago but didn't make any progress on it. (There was some oddity with detecting when just the template part block was selected that I didn't figure out.) A new PR could be opened once someone wants to pick this up again 👍
